Product Collection: Return to top after page change

imanish003 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR introduces a "Scroll to Top" feature upon page change. It also involves a significant refactor in how we utilize the Interactivity API for client-side pagination i.e. the transition from using the `navigation-link` directive to `actions`.

Closes #42195 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new post & Add Product Collection block
2. Select “Product Catalog” collection
3. Publish the changes & go to Frontend.
4. After changing pages, confirm that the page scrolls to the top of the Product Collection block, as illustrated in the GIF below:
![Screen Capture on 2024-01-23 at 12-21-57 pm](https://github.com/woocommerce/woocommerce/assets/16707866/49c38891-818b-42a4-b528-75b8fd4e0624)

5. In the Editor, adjust the “ITEMS PER PAGE” setting in the toolbar controls to 3.
6. Save these changes and ensure that pagination functions correctly on the Frontend.
    1. Experiment with changing pages rapidly and check that the pagination block appears and operates as expected.
7. Now add multiple Product Collection to same post & verify that:
    1. Pagination works as expected.
    2. Confirm that the "Scroll to top" feature operates properly for the Product Collection block when you change its page.

Here is a quick demo video:

https://github.com/woocommerce/woocommerce/assets/16707866/099326c0-7ea6-447d-b41b-bddac470cbe3

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Product Collection: Return to top after the page change

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

imanish003 committed:
Initial working version

imanish003 committed:
Update scrolling behavior in product collection block frontend

This commit updates the scrolling behavior in the product collection block's frontend script.

Previously, the code focused on the first anchor or button within a product collection block. This approach was taken to maintain accessibility and ensure that the focused element was scrolled into view. However, this behavior has been changed to improve the user experience.

The updated code now directly scrolls to the first product in the collection. This is achieved by selecting the product using a new selector that targets the `.wc-block-product` class within the `.wc-block-product-template` of the specified `data-wc-navigation-id`. Once the product is selected, the `scrollIntoView` method is used with smooth behavior and start block alignment, offering a more visually appealing scroll effect.

This change enhances the user experience by smoothly directing attention to the beginning of the product collection, making it easier for users to browse products.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/42195-return-to-top-behavior-when-loading-new-page

imanish003 committed:
Remove code related to animation

imanish003 committed:
Fix the issue related to pagination block rendering

- I have improved the code by breaking it into smaller functions
- Added unique `data-wc-key` to each anchor tag

imanish003 committed:
Enhance Product Collection Block with Interactive Prefetching

imanish003 committed:
Refine Prefetch Logic & other improvements

This commit further refines the prefetch logic and interactivity in the Product Collection block. It focuses on enhancing code clarity and improving the prefetch behavior based on user interactions.

Key changes include:
1. Renamed 'isPrefetchNextAndPreviousLink' to 'isPrefetchNextOrPreviousLink' in ProductCollectionStoreContext for better readability and accuracy.
2. Modified the scrollToFirstProductIfNotVisible function to accept a wcNavigationId as a parameter instead of a reference
3. Introduced checks for the existence of wcNavigationId in scrollToFirstProductIfNotVisible to prevent unnecessary executions.
4. Updated comments throughout frontend.tsx for enhanced clarity, explaining the prefetch logic and its triggers in detail.
5. Adjusted the logic in the navigate and prefetch functions to align with the new context structure and prefetch strategy.
6. In ProductCollection.php, updated data attributes to align with the revised prefetch strategy.
7. Refined the attribute setting in process_pagination_links function to correctly implement the updated interactivity and prefetch logic.

These changes aim to optimize the prefetch functionality, reduce unnecessary prefetching on initial page load, and ensure a smoother and more efficient navigation experience within the Product Collection block.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 committed:
Improve comments

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -122,7 +113,16 @@ public function add_navigation_id_directive( $block_content, $block, $instance )
 					'wc-product-collection-' . $this->parsed_block['attrs']['queryId']
 				);
-				$p->set_attribute( 'data-wc-interactive', wp_json_encode( array( 'namespace' => 'woocommerce/product-collection' ) ) );
+				$p->set_attribute( 'data-wc-interactive', '{ "namespace": "woocommerce/product-collection" }' );
```
Can you explain why we don't need that encoding anymore?

imanish003 replied:
I aimed to maintain consistency with the syntax used in the Query Loop block, as you can observe [here](https://github.com/WordPress/gutenberg/blob/621a3cd0ee1b86e0a213e820a159fac033cb7a7b/packages/block-library/src/query/index.php#L30). Do you foresee any issues with this approach?

CC: @DAreRodz, Was there a specific reason for the syntax in the query loop block to be different?

imanish003 replied:
Also, after implementing that change, everything is functioning as anticipated.

DAreRodz replied:
There is no difference between any of the approaches; both render the same JSON code. So, there's no specific reason other than avoiding potential mistakes while manually writing a JSON data structure. 😄

kmanijak replied:
> Do you foresee any issues with this approach?

No, like I wrote in main comment, the questions "are only for my education and do not intend to challenge anything." 😆 I have trust in what's happening in Query Loop and your assessment so don't worry too much about my questions!

imanish003 replied:
> there's no specific reason other than avoiding potential mistakes while manually writing a JSON data structure

Understood. Thanks for the quick response. I will revert the change. 😀

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -132,49 +132,72 @@ public function add_navigation_id_directive( $block_content, $block, $instance )
+				'class_name' => $class_name,
+			)
+		) ) {
+			$processor->set_attribute( 'data-wc-interactive', '{ "namespace": "woocommerce/product-collection" }' );
```
Isn't this a duplication of what we do in `add_navigation_id_directive` method https://github.com/woocommerce/woocommerce/pull/43817/files#diff-933825fdae3edb1301562e70acef6627856dd1c39c7475056ef21ac4f0ee9941R116? Or the other way around, duplicating there what we do in here?

imanish003 replied:
The reason was that if we don’t explicitly add a namespace using the following line:

```jsx
$processor->set_attribute( 'data-wc-interactive', '{ "namespace": "woocommerce/product-collection" }' );
```

We need to explicitly add namespace with each callback & action, for example:

```jsx
$processor->set_attribute( 'data-wc-watch', 'woocommerce/product-collection::callbacks.prefetch' );
```

@DAreRodz @luisherranz Is there any reason why we need to explicitly add namespace with actions and callbacks if parent (Product Collection) block have defined namespace? AFAIU It should automatically inherit namespace from it’s parent 🤔

DAreRodz replied:
When I decided to include the namespace explicitly on the directive values, my reasoning was that, as they were placed in a different block, another interactive block could be placed in between and overwrite the namespace, thus making the `data-wc-watch` directives look for the `callbacks.prefetch` in the wrong namespace.

Another way to prevent this would be to add a `data-wc-interactive` directive with the required namespace, but I wanted to try the other syntax. 🙂

imanish003 replied:
> another interactive block could be placed in between and overwrite the namespace


For me, there was no interactive block in-between. There was just Product Collection block. I am not sure why following code wasn’t working:

```jsx
$processor->set_attribute( 'data-wc-watch', 'callbacks.prefetch' );
```

If there is a parent block (Product Collection) with namespace `woocommerce/product-collection`, shouldn’t it use “callbacks.prefetch” from this namespace? I tried it, and it wasn’t working.

DAreRodz replied:
> shouldn’t it use “callbacks.prefetch” from this namespace?

It should! That's interesting. Could you check that there isn't any `data-wc-interactive` in between with a different namespace in the final HTML? That's the only thing I can think of that could break it. 🤔

imanish003 replied:
@DAreRodz, I've just recorded a video to explain the issue, as I thought it would be more effective than a written explanation. Please correct me if I misunderstood how namespace should be working 🙂

https://github.com/woocommerce/woocommerce/assets/16707866/074090ec-10ee-443f-9936-e20ff4f129b2

kmanijak reviewed (approved):
Everything works just fine! Thanks for taking care of it! 🚀 

I left some questions but they are only for my education and do not intend to challenge anything.

imanish003 committed:
Use wp_json_encode for data-wc-interactive attribute

This commit updates the `ProductCollection` class to use `wp_json_encode` for setting the `data-wc-interactive` attribute value. Previously, the attribute value was hardcoded as a JSON string. Now, by utilizing `wp_json_encode`, we ensure the JSON encoding is handled correctly by WordPress standards, enhancing readability and maintainability of the code.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/42195-return-to-top-behavior-when-loading-new-page
